proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -3026,19 +3021,758 @@ lemma reply_unlink_ccorres:
+  apply simp
+  done
+
+(* FIXME RT: improve naming of this and other ccorres lemmas for updateSchedContext *)
+lemma updateSchedContext_ccorres_lemma:
+  assumes spec: "\<forall>s. \<Gamma>\<turnstile> \<lbrace>s. P s\<rbrace> Call f {t. Q s t}"
+  and      mod: "modifies_heap_spec f"
+  and  rl: "\<And>s s' t ko. \<lbrakk>(s, s') \<in> rf_sr; Q s' t; s' \<in> P'; ko_at' ko scPtr s\<rbrakk>
+             \<Longrightarrow> (s\<lparr>ksPSpace := (ksPSpace s)(scPtr \<mapsto> KOSchedContext (g ko))\<rparr>,
+                  t\<lparr>globals := globals s'\<lparr>t_hrs_' := t_hrs_' (globals t)\<rparr>\<rparr>) \<in> rf_sr"
+  and  g:  "\<And>s s'. \<lbrakk>sc_at' scPtr s; s' \<in> P'; (s, s') \<in> rf_sr\<rbrakk> \<Longrightarrow> P s'"
+  and h: "\<And>ko. scSize (g ko) = scSize ko"
+  shows "ccorres dc xfdc (sc_at' scPtr and obj_at' (\<lambda>ko. objBits ko < word_bits) p) P' hs
+           (updateSchedContext scPtr g) (Call f)"
+  apply (rule ccorres_Call_call_for_vcg)
+   apply (rule ccorres_from_vcg)
+   apply (rule allI, rule conseqPre)
+   apply (rule HoarePartial.ProcModifyReturnNoAbr
+                [where return' = "\<lambda>s t. t\<lparr> globals := globals s\<lparr>t_hrs_' := t_hrs_' (globals t) \<rparr>\<rparr>"])
+     apply (rule HoarePartial.ProcSpecNoAbrupt[OF _ _ spec])
+      apply (rule subset_refl)
+     apply vcg
+    prefer 2
+    apply (rule mod)
+   apply (clarsimp simp: mex_def meq_def)
+  apply clarsimp
+  apply (rule conjI)
+   apply (erule (2) g)
+  apply clarsimp
+   apply (frule obj_at_ko_at')
+   apply clarsimp
+  apply (prop_tac "objBits ko < word_bits")
+   apply (clarsimp simp: obj_at_simps)
+  apply (rule bexI [rotated])
+   apply (erule updateSchedContext_eq)
+    apply (insert h)
+    apply simp
+   apply (clarsimp simp: obj_at_simps)
+  apply simp
+  apply (rule_tac x1="t\<lparr>globals := globals x\<lparr>t_hrs_' := t_hrs_' (globals t)\<rparr>\<rparr>"
+               in iffD1 [OF rf_sr_upd], simp_all)[1]
+  apply (erule (3) rl)
+  done
+
+lemma updateSchedContext_ccorres_lemma4:
+  "\<lbrakk> \<And>s sc. \<Gamma> \<turnstile> (Q s sc) c {s'. (s \<lparr>ksPSpace := (ksPSpace s)(scPtr \<mapsto> injectKOS (g sc))\<rparr>, s') \<in> rf_sr};
+     \<And>sc. scSize (g sc) = scSize sc;
+          \<And>s s' sc sc'. \<lbrakk> (s, s') \<in> rf_sr; P sc; ko_at' sc scPtr s;
+                             cslift s' (Ptr scPtr) = Some sc';
+                             csched_context_relation sc sc'; P' s ; s' \<in> R\<rbrakk> \<Longrightarrow> s' \<in> Q s sc \<rbrakk>
+   \<Longrightarrow> ccorres dc xfdc
+         (obj_at' (P :: sched_context \<Rightarrow> bool) scPtr and P') R hs
+         (updateSchedContext scPtr g) c"
+  apply (rule ccorres_from_vcg)
+  apply (rule allI)
+  apply (case_tac "obj_at' P scPtr \<sigma>")
+   apply (drule obj_at_ko_at', clarsimp)
+   apply (rule conseqPre, rule conseqPost)
+      apply assumption
+     apply clarsimp
+     apply (rule rev_bexI, rule updateSchedContext_eq)
```

Comment:
clarsimp should discharge the assumption, then we have clarsimp, three simps, and a clarsimp... I think you can safely put a `; clarsimp simp: obj_at_simps` at the end here and save some lines

Reply:
(or clarsimp with a +)

Reply:
These were just moved. I meant to write a nicer PR description before I marked this as ready for review. But these `updateSchedContext_ccorres_*` lemmas were just moved, along with `schedContext_donate_ccorres` and `tcbReleaseRemove_ccorres`

Reply:
It's still a pretty easy compression if you're interested while we're here.